Support HTTP persistent connections in Web Cmdlets

stevenebutler opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Fis #12764

This PR addresses [12764](https://github.com/PowerShell/PowerShell/issues/12764) by using persistent HTTP connections when Invoke-WebRequest uses the -Session variable. It does this by 
1. Tracking the  HttpClient object in the saved WebRequestSession.
2. Monitoring properties in WebRequestSession and rebuilding the HttpClient if they are changed. This is done to ensure that existing scripts that might tweak parameters between invocations of IWR actually apply the changes so they will behave in a backwards compatible manner. When anything that impacts the HttpClient is changed, the persistent connection is abandoned by disposing the HttpClient and underlying HttpClientHandler and constructing a new one. A warning is logged when this occurs as it has a performance impact (though no worse than the existing behaviour).
3. To support point 2, many parameters are stored as internal set-only properties in the WebRequestSession so changes between invocations that would require a new connection can be detected and the handler rebuilt appropriately.

## PR Context

This PR makes a considerable difference to performance when using Invoke-WebRequest or Invoke-RestMethod with sessions when multiple requests are made to the same host. This is particularly evident for HTTPS requests, as each HTTPS connection can take hundreds of milliseconds to establish.

This PR as written passes existing tests in `WebCmdlets.Tests.ps1` and additional WebSession tests that verify the HttpClient is rebuilt when appropriate.

The PR provides as much backwards compatibility as possible at the expense of the complexity of tracking how the HttpClient used in the WebSession was constructed and recreating the HttpClient in the case it would have been constructed differently between invocations in scripts that assume it is okay to change parameters between invocations with the same WebSession.

This PR squashes the commits in the original PR#19173 following multiple round of reviews by @iSazonov and @CarloToso on a previous version of the PR. PR#19173 has been withdrawn due to number of commits and volume of review comments.

## Performance Impact
Performance impact of maintaining HTTP connection persistence is measured by comparing PowerShell 7.3.3 with this PR using this simple script:

```powershell
(Measure-Command {
    $null =Invoke-WebRequest https://postman-echo.com/get?i=1 -SessionVariable Session
    1..10 | ForEach-Object {
            $null =Invoke-WebRequest https://postman-echo.com/get?i=$_ -WebSession $Session
    }
}).ToString()
```

### PowerShell 7.3.3
> 00:00:08.3019013

### PR Branch
> 00:00:03.4589770

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

stevenebutler commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -894,29 +935,55 @@ internal virtual void PrepareSession()
+                // SslProtocol parameter is a value type so we only want to switch back to the default
+                // if it is explicitly provided on the command line. Otherwise we keep the value in
+                // the WebSession from any previous invocation.
+                WebSession.SslProtocol = SslProtocol;
```
Comment was added to explain why it is being careful not to blindly set the WebSession.SslProtocol to SslProtocol parameter. Evidence this is required is that one of the review comments said I should just set it directly.

iSazonov replied:
If this comment only says that if SslProtocol is present, its value should be stored in WebSession.SslProtocol = SslProtocol, then this is an obvious comment and I would delete it.

CarloToso replied:
He added the comment because I had misunderstood the idea behind the code https://github.com/PowerShell/PowerShell/pull/19173#discussion_r1111123122

stevenebutler replied:
I have removed the comment.

stevenebutler commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -894,29 +935,55 @@ internal virtual void PrepareSession()
+                        webProxy.UseDefaultCredentials = true;
+                    }
+
+                    // We don't want to update the WebSession unless the proxies are different
+                    // as that will require us to create a new HttpClientHandler and lose connection
+                    // persistence.
+                    if (!webProxy.Equals(WebSession.Proxy))
+                    {
+                        WebSession.Proxy = webProxy;
+                    }
```
WebRequestSession does not do the check internally because it is using shared code to check for object equality when setting the cookies, proxy, certificates etc. These are publically settable properties on WebSession so can be changed directly in script. In general, these must be updated if the object references aren't the same as the user can modify the properties of the referenced objects after they are set on the WebSession. The WebProxy class is internal so can't easily be constructed by the user, and is a type used directly only by the WebRequestPSCmdlet so it makes sense to add the special case optimisation for not updating the internal proxy class if the cmdlet proxy parameters are identical from invocation to invocation in the class that directly knows about the equality optimisation.

The alternative if to use .Equals() for all reference types in the SetClassVar() function in WebRequestSession, which would be problematic if the equality operator says different references are Equal, but the user tries to modify an object after setting it in the WebRequestSession and it is ignored because of the Equality check.

iSazonov replied:
I agree to keep this as-is.
Not clear how underlying .Net API responds to changes contents of reference object parameters.

stevenebutler commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -4,15 +4,38 @@
+        /// Record whether an existing HttpClient in the WebRequestSession had to be disposed
+        /// after creating due to requested property changes.
+        /// </summary>
+        private bool _disposedClient;
```
done.

stevenebutler commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -105,8 +158,141 @@ public WebRequestSession()
+            // Set the credentials used by this request
+            if (UseDefaultCredentials)
+            {
+                // The UseDefaultCredentials flag overrides other supplied credentials
```
I don't understand what you mean here. Are you referring to the PR being worked on by @CarloToso ?

iSazonov replied:
`if (Credentials is not null)` must be before `if (UseDefaultCredentials)`
If we remove `else if` in line 204 we create a bug.

stevenebutler replied:
Is this something being worked on in another PR? Should I wait for that to merge before applying here?

stevenebutler replied:
Think this was part of the changes I merged recently. Please check again.

stevenebutler commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -366,6 +366,14 @@ function GetMultipartBody {
 }
 
+filter FilterVerboseMessages ([string]$Message) {
+    if ($_ -and $_ -is [System.Management.Automation.VerboseRecord]) {
```
I looked into the .Streams.Verbose and had a lot of trouble with it, but have something that works now. There was no shared code, but plenty of places with examples of how it was done.

I think my use case is a bit different because I am wanting to run multiple commands in the pipeline with state between each and then check whether a warning is raised each time. I have to move the state in and out of the created PowerShell session. 

In the context for persistence checks I have this function:
```powershell

    $connWarning = [regex]::new('WebSession.+HTTP')
    function RunCheckingPersistence {
            param(
                [uri]$Uri,
                [string]$Command,
                [object]$Session,
                [switch]$LostPersistence,
                [switch]$CaptureSession
            )

            $pwsh = [PowerShell]::Create()
            $pwsh.Runspace.SessionStateProxy.SetVariable('uri', $Uri)
            if ($Session) {
                $pwsh.Runspace.SessionStateProxy.SetVariable('Session', $Session)
                $command = "$command -WebSession `$Session"
            }
            if ($CaptureSession) {
                $command = "$command -SessionVariable Session"
            }
            $script = "`$null = $command -Verbose"
            $pwsh.AddScript($script).Invoke()
            $session = $pwsh.Runspace.SessionStateProxy.GetVariable('Session')

            $expectedWarnings = if ($LostPersistence) { 1 } else { 0 }
            ($pwsh.Streams.Verbose | Where-Object { $connWarning.Matches($_.Message) }).Count | Should -Be $expectedWarnings

            $pwsh.Dispose()

            return $session
        }
```

And use like this:
```powershell
  It 'Connection persistence impacted by changing SkipCertificateCheck' {
            $uri = Get-WebListenerUrl -Https
            # This first request will throw because the certificate is invalid
            $session = $null
            $session = RunCheckingPersistence -Uri $uri  -Command 'Invoke-WebRequest -Uri $uri -SkipCertificateCheck:$false' -CaptureSession
            # No change in setting
            $session = RunCheckingPersistence -Uri $uri  -Command 'Invoke-WebRequest -Uri $uri -SkipCertificateCheck:$false' -Session $session
            # Skipping cert check changes persistence
            $session = RunCheckingPersistence -Uri $uri  -Command 'Invoke-WebRequest -Uri $uri -SkipCertificateCheck' -Session $session -LostPersistence
            # Same settings won't lost persistence
            $session = RunCheckingPersistence -Uri $uri  -Command 'Invoke-WebRequest -Uri $uri -SkipCertificateCheck' -Session $session
            # Lose persistence due to changing cert check - this will also throw
            $session = RunCheckingPersistence -Uri $uri  -Command 'Invoke-WebRequest -Uri $uri -SkipCertificateCheck:$false' -Session $session -LostPersistence
        }
```
Before I commit all these changes, could you please give me some feedback on the approach. This took me a long time to get working because I couldn't just keep adding commands to the $pwsh item without it running all commands every time I ran it, and when I used the Clear commands as below between invocations of the same $pwsh, it stopped working for subsequent requests and gave me nothing in the stream.

```powershell
            $pwsh.Commands.Clear()
            $pwsh.Streams.ClearStreams()
```

iSazonov replied:
Perhaps we could makes the test more simple with test hook. See `InternalTestHooks` in Utils.cs.

stevenebutler replied:
I'm not familiar with that, but it looks like it's used to control behaviour with flags while testing?

Are you suggesting I add a boolean to the static class there with something like:

```csharp
        //True if the last Invoke-WebRequest or Invoke-RestMethod had to reset the connection
        internal static bool LastWebRequestResetConnection;
```

And then gain access to the boolean after each request to validate?

stevenebutler replied:
In the meantime, I have committed the tests using the Verbose stream approach.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2201,178 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+        It 'Connection persistence outperforms non-persistence' {
+            $uri = Get-WebListenerUrl -Https
+            $null = Invoke-WebRequest -Uri $uri -SessionVariable Session -SkipCertificateCheck
+            $iters = 5
+            $persistence = Measure-Command {
+                1 .. $iters | ForEach-Object {
```
I am ok to remove the test. Only please add perf comparisons in the PR description.

stevenebutler replied:
Ok, the tests are very fast when running with the mock service, since it's all localhost. The connection setup, even for SSL is pretty quick also, since we have to ignore certificate errors. A comparison with a real HTTPS site with certificate checking gives much bigger performance improvement from persistence.

I will update the PR with some benchmarks against the postman site as per the issue.

stevenebutler commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2201,178 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+            $persistence | Should -BeLessThan $nonPersistence
+        }
+
+        It 'Disable cert checking alters persistence' {
```
Updated these, but waiting on feedback about the hooks before I commit anything.

stevenebutler commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2193,6 +2201,178 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+            $verboseMessages | Should -BeNullOrEmpty
+        }
+
+        It 'Proxy can stay off between tests' {
```
Updated

stevenebutler commented:
Codefactor is complaining about some tests that were not added by this PR using plain text as a password.

@CarloToso will that impact being able to merge this PR?

CarloToso commented:
@stevenebutler Don't worry about those two codefactor issues, I introduced the in #19190

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -105,8 +157,140 @@ public WebRequestSession()
+            handler.AutomaticDecompression = DecompressionMethods.All;
+
+            // Set the credentials used by this request
+            if (Credentials is not null)
```
Please remove obvious comment.

stevenebutler replied:
Removed.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2236,6 +2236,146 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+            $session = $pwsh.Runspace.SessionStateProxy.GetVariable('Session')
+
+            $expectedWarnings = if ($LostPersistence) { 1 } else { 0 }
+            ($pwsh.Streams.Verbose | Where-Object { $connWarning.Matches($_.Message) }).Count | Should -Be $expectedWarnings
```
Please update the variable name - it is not for `warning`.

stevenebutler replied:
Changed variables to reflect the message better:
* `$expectedConnRecreatedCount` for the 0 or 1 count of messages
* `$matchConnRecreatedMessage` for the regex
* `-ExpectConnectionRecreated` for the function parameter

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2236,6 +2236,146 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
     }
+
+    Context 'Invoke-WebSession: Connection persistence in a WebSession' {
+        $connWarning = [regex]::new('WebSession.+HTTP')
```
I'd add a comment with resource string name and its content for what message we expect.

stevenebutler replied:
Done. Also updated variable name to better match the resource string.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2236,6 +2236,146 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+                [uri]$Uri,
+                [string]$Command,
+                [object]$Session,
+                [switch]$LostPersistence,
```
Maybe 
```suggestion
                [switch]$ExpectLostPersistence,
```

stevenebutler replied:
Used `-ExpectConnectionRecreated` for the function parameter to better match the resource string name

stevenebutler committed:
Support HTTP persistent connections in Web Cmdlets

Use persistent HTTP connections when Invoke-WebRequest uses the -Session variable.

This is achieved by
* tracking the HttpClient object in the saved WebRequestSession
* monitoring properties in WebRequestSession and rebuilding the
  HttpClient if they are changed. This is done to ensure that
  existing scripts that might change parameters between invocations
  of IWR actually apply the changes so they will behave in a backwards
  compatible manner. When anything that impacts the HttpClient is
  changed, the persistent connection is abandoned by disposing the
  HttpClient and underlying HttpClientHandler and constructing a new
  one. A verbose message is logged when this occurs as it has a
  performance impact (though no worse than the existing behaviour).
* To support the point 2 above, many switches were added as internal
  set-only properties to the WebRequestSession so changes between
  invocations can be detected and the handler rebuilt appropriately.

stevenebutler commented:
I have squashed all commits into a single commit.

adityapatwardhan commented:
@stevenebutler Thank you for your contribution!

tonyg00 commented:
I tested Preview 2 and back to back invocations of Invoke-WebRequest and Invoke-RestMethod are not re-using ports, new ones are created.

CarloToso commented:
@tonyg00 did you use the `-SessionVariable` and then `-WebSession` parameter?

tonyg00 commented:
I did not, I just executed, Invoke-WebRequest -uri <host/uri> which is the most common use case for me of these cmdlets.
What is the correct syntax for this?

CarloToso commented:
Don't worry, it's a new feature, it's not yet documented.
I think this syntax should work:
```ps1
Invoke-WebRequest -Uri -SessionVariable session
Invoke-WebRequest -Uri -WebSession $session
```
